functions/database/database.js

```diff
@@ -0,0 +1,316 @@
+  const today = Timestamp.now();
+  const releasesRef = db.collection("releases");
+  const query = releasesRef.where("codeFreezeDate", ">", today);
+  const querySnapshot = await query.get();
+  const batch = db.batch();
+
+  querySnapshot.forEach((doc) => {
+    batch.delete(doc.ref);
+  });
+
+  await batch.commit();
+}
+
+/**
+   * Update a release in Firestore.
+   *
+   * @param {string} releaseId The ID of the release to update.
+   * @param {string} releaseData The new data to set for the release.
+   * @throws {Error} If the release state update fails.
+   */
+async function updateRelease(releaseId, releaseData) {
+  validateRelease(releaseData);
+  const releaseDoc = db.collection("releases").doc(releaseId);
+  await releaseDoc.update(releaseData);
+}
+
+/**
+ * Updates the state of a release in the database.
+ *
+ * @param {string} releaseId The ID of the release to update
+ * @param {string} newState The new state to set for the release
+ * @throws {Error} If the release state update fails, or if the
+ * release state is invalid.
+ */
+async function updateReleaseState(releaseId, newState) {
+  // Verify that the state is one of the valid states
+  if (!Object.values(RELEASE_STATES).includes(newState)) {
+    throw new Error(`Invalid release state: ${newState}`);
+  }
+
+  const releaseDoc = db.collection("releases").doc(releaseId);
+  await releaseDoc.update({state: newState});
+}
+
+/**
+ * Creates new library release documents for each version in the libraryVersions
+ * object, and deletes any existing library versions associated with the
+ * release.
+ *
+ * @param {Object} libraryVersions Object mapping library names to their
+ * versions.
+ * @param {string} releaseId The ID of the associated release.
+ */
+async function updateLibrariesForRelease(libraryVersions, releaseId) {
+  const batch = db.batch();
+
+  // Delete all previous library versions for this release
+  const previousLibrariesSnapshot = await db.collection("libraries")
+      .where("releaseID", "==", releaseId)
+      .get();
+  previousLibrariesSnapshot.docs.forEach((doc) => {
+    const docRef = db.collection("libraries").doc(doc.id);
+    batch.delete(docRef);
+  });
+
+  // Write each library version to Firestore
+  Object.entries(libraryVersions).forEach(([libraryName, updatedVersion]) => {
+    const docRef = db.collection("libraries").doc();
+    batch.set(docRef, {
+      libraryName,
+      updatedVersion,
+      releaseID: releaseId,
+    });
+  });
+
+  await batch.commit();
+}
+
+
+/**
+ * Creates new change documents for each change in the release report, and
+ * deletes any existing changes associated with the release.
+ *
+ * @param {Object} releaseReport The release report containing changes by
+ * library name.
+ * @param {string} releaseId The ID of the associated release.
+ * @throws {Error} If a library in the release report does not exist in
+ * Firestore.
+ */
+async function updateChangesForRelease(releaseReport, releaseId) {
```

Comment:
Ideally, this function would be like:
```js
async function updateChangesForRelease(releaseReport, releaseId) {
  const batch = db.batch();

  await deletePreviousChangesForID(batch, releaseId);
  await releaseReport.changesByLibraryName.map((library) => createChange(batch, library));
  
  await batch.commit();
}
```

Or something of the like, if you get my drift.

Reply:
Resolved in https://github.com/firebase/firebase-release-dashboard/pull/8/commits/030399c9ed011642c29c1d0d1c946d65c8e3d17f.